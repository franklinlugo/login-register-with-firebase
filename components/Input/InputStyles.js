import styled, { keyframes, css } from 'styled-components';

export const StyledInput = styled.input`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 0 0 0 32px;
  color: ${(props) => props.theme.gray};
  border-radius: 3px;
  border: none;
  z-index: 1000;
  outline: none;
`;

export const StyledLabel = styled.label`
  position: absolute;
  top: 50%;
  left: 32px;
  padding: 4px 6px;
  background-color: #f7f7f7;
  transform: translate(0, -50%);
  line-height: 1;
  transition: 0.3s;
  border-radius: 3px;
  color: ${(props) => props.theme.gray};
`;

export const IconWrapper = styled.div`
  position: absolute;
  top: 50%;
  left: 8px;
  transform: translate(0, -50%);
  width: 28px;
  transition: 0.3s;
  svg {
    path {
      fill: #7f7f7f;
    }
  }
`;

const shadowAnimation = keyframes`
  to {
    box-shadow: 0px 0px 60px 20px;
    opacity: 0;
  }
`;

export const Focus = styled.div`
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  box-shadow: 0px 0px 0px 0px;
  border-radius: 3px;
  color: transparent;
`;

const wrapperFocusedMixin = css`
  border-color: ${(props) => props.theme.primary};
  ${StyledLabel} {
    top: 0;
    left: 26px;
  }
  ${Focus} {
    animation: ${shadowAnimation} 0.5s ease-in-out forwards;
    color: rgba(211, 63, 141, 0.6);
  }
  ${IconWrapper} {
    svg {
      path {
        fill: ${(props) => props.theme.primary};
      }
    }
  }
`;

const wrapperWithValueMixin = css`
  ${StyledLabel} {
    top: 0;
    left: 26px;
  }
`;

export const InputContainer = styled.div`
  position: relative;
  width: 100%;
  height: 60px;
  /* margin-bottom: 24px; */
  border-radius: 3px;
  background-color: #f7f7f7;
  border: 1px solid;
  border-color: transparent;
  transition: 0.3s;
  ${({ isFocused }) => isFocused && wrapperFocusedMixin}
  ${({ isEmpty }) => !isEmpty && wrapperWithValueMixin}
`;

export const Error = styled.p`
  display: grid;
  grid-template-columns: min-content 1fr;
  align-items: center;
  gap: 8px;
  margin: 4px 0 0 0;
  padding: 0;
  line-height: 1;
  color: #c80000;
`;

export const Container = styled.div`
  margin-bottom: 24px;
`;
