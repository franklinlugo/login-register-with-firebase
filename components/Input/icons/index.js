export { default as Envelope } from './Envelope';
export { default as Lock } from './Lock';
