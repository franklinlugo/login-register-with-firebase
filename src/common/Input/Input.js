import React, { useState, useEffect, useRef } from 'react';
import { string, oneOf, func, bool } from 'prop-types';
import { Envelope, Lock, Error as ErrorIcon, EyeOpen, EyeClosed } from '../../assets/Icons';
import {
  Container,
  InputContainer,
  StyledLabel,
  StyledInput,
  IconWrapper,
  Focus,
  Error,
  EyeVisibility,
} from './InputStyles';

const IconEnum = Object.freeze({
  envelope: Envelope,
  lock: Lock,
});

function Input({ name, id, type = 'text', label, value = '', onChange, icon, showError, errorMessage }) {
  const [isFocused, setIsfocused] = useState(false);
  const [isEmpty, setIsEmpty] = useState(true);
  const isPassword = useInputIsPasswordType(type);

  useEffect(() => {
    setIsEmpty(value.toString().length === 0);
  }, [value]);

  const handleOnFocus = () => setIsfocused(true);

  const handleOnBlur = () => setIsfocused(false);

  const [isPasswordVisible, setIsPasswordVisible] = useState(false);
  const [derivedInputType, setDerivedInputType] = useState('password');
  const handlePasswordVisibility = () => {
    setIsPasswordVisible((_) => !_);
    setDerivedInputType((_) => (_ === 'password' ? 'text' : 'password'));
  };

  const Icon = IconEnum[icon];
  return (
    <Container>
      <InputContainer isFocused={isFocused} isEmpty={isEmpty}>
        <StyledLabel htmlFor={id}>{label}</StyledLabel>
        <StyledInput
          name={name}
          id={id}
          type={isPassword ? derivedInputType : type}
          value={value}
          onChange={onChange}
          onFocus={handleOnFocus}
          onBlur={handleOnBlur}
        />
        <IconWrapper>
          <Icon width="18px" />
        </IconWrapper>
        {isPassword && (
          <EyeVisibility onClick={handlePasswordVisibility} role="button">
            {isPasswordVisible ? <EyeOpen /> : <EyeClosed />}
          </EyeVisibility>
        )}
        <Focus />
      </InputContainer>
      {showError && (
        <Error>
          <ErrorIcon width="12px" />
          {errorMessage}
        </Error>
      )}
    </Container>
  );
}

Input.propTypes = {
  name: string,
  id: string,
  type: string,
  label: string,
  value: string,
  onChange: func,
  icon: oneOf(['envelope', 'lock']),
  showError: bool,
  errorMessage: string,
};

function useInputIsPasswordType(type) {
  const isPassword = useRef(type === 'password');
  return isPassword.current;
}

export default Input;
