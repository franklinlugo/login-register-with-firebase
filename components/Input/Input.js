import React, { useState, useEffect } from 'react';
import { string, oneOf, func, bool } from 'prop-types';
import { Envelope, Lock, Error as ErrorIcon } from '../../assets/Icons';
import { Container, InputContainer, StyledLabel, StyledInput, IconWrapper, Focus, Error } from './InputStyles';

const IconEnum = Object.freeze({
  envelope: Envelope,
  lock: Lock,
});

function Input({ name, id, type = 'text', label, value = '', onChange, icon, showError, errorMessage }) {
  const [isFocused, setIsfocused] = useState(false);
  const [isEmpty, setIsEmpty] = useState(true);

  useEffect(() => {
    setIsEmpty(value.toString().length === 0);
  }, [value]);

  const handleOnFocus = () => setIsfocused(true);

  const handleOnBlur = () => setIsfocused(false);

  const Icon = IconEnum[icon];
  return (
    <Container>
      <InputContainer isFocused={isFocused} isEmpty={isEmpty}>
        <StyledLabel htmlFor={id}>{label}</StyledLabel>
        <StyledInput
          name={name}
          id={id}
          type={type}
          value={value}
          onChange={onChange}
          onFocus={handleOnFocus}
          onBlur={handleOnBlur}
        />
        <IconWrapper>
          <Icon width="18px" />
        </IconWrapper>
        <Focus />
      </InputContainer>
      {showError && (
        <Error>
          <ErrorIcon width="12px" />
          {errorMessage}
        </Error>
      )}
    </Container>
  );
}

Input.propTypes = {
  name: string,
  id: string,
  type: string,
  label: string,
  value: string,
  onChange: func,
  icon: oneOf(['envelope', 'lock']),
  showError: bool,
  errorMessage: string,
};

export default Input;
