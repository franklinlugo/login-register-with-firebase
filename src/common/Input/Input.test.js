import React from 'react';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from './Input';

test('Input render correctly', () => {
  const onChangeMock = jest.fn();
  const { asFragment } = render(<Input name="input" id="input" label="label" value="value" onChange={onChangeMock} />);

  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByRole('textbox')).toHaveValue('value');

  userEvent.type(screen.getByRole('textbox'), '123');
  expect(onChangeMock).toBeCalledTimes(3);
});

test('Input handle password', () => {
  render(<Input type="password" id="input" label="label" value="value" />);

  expect(screen.getByLabelText('label')).toHaveAttribute('type', 'password');

  userEvent.click(screen.getByRole('button'));
  expect(screen.getByLabelText('label')).toHaveAttribute('type', 'text');
});

test('Input handle error', () => {
  const errorMessage = 'error message';
  render(<Input id="input" label="label" value="value" showError errorMessage={errorMessage} />);

  expect(screen.getByText(errorMessage)).toBeInTheDocument();
});
