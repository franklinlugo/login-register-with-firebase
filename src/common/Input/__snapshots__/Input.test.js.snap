// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-Axmtr hPYRYX"
  >
    <div
      class="sc-AxgMl jLLscj"
    >
      <label
        class="sc-AxirZ dvsQlS"
        for="input"
      >
        label
      </label>
      <input
        class="sc-AxjAm kBvMjT"
        id="input"
        name="input"
        type="text"
        value="value"
      />
      <div
        class="sc-AxhUy bMCNhl"
      />
    </div>
  </div>
</DocumentFragment>
`;
